### PR TITLE
Add metrics/alloc_bytes for showing memory allocated

### PR DESCRIPTION
**Pull request by anderseknert**
Add a simple endpoint (`/metrics/alloc`) to quickly show memory utilization of OPA. This can be done via e.g. Prometheus already, but having a human friendlier interface for this is nice. 

```shell
$ opa run -s https://bundles.example.com/my-big-bundle.tar.gz
```
```shell
$ curl 'localhost:8181/metrics/alloc?pretty=true'
3.3GB
```

Without the `pretty` query string, the raw number of bytes is shown. I'm considering whether the pretty format should be the default, and something like `format=raw`, or `pretty=false` should be the option to show bytes, as that's likely less useful to a human operator. Ideas on that would be appreciated.

NOTE: There's a slight discrepancy between what Go reports and what top and other programs do (Go reports less memory used), but I think this is good enough for providing a ballpark metric.

### CODE REVIEW

**Review by srenatus — DISMISSED** [review]
Code looks good. My only concern would be that the same thing could be achieved by reading the existing metrics and doing the addition and pretty printing using some favourite scripting language. But it surely adds some convenience over that.

**Comment by srenatus** [comment]
> NOTE: There's a slight discrepancy between what Go reports and what top and other programs do (Go reports less memory used), but I think this is good enough for providing a ballpark metric.

I haven't read the detailed descriptions of the metrics added together here, but if anything, the runtime metrics should be more accurate over top etc, shouldn't they be?

**Comment by anderseknert** [comment]
You're absolutely right: this is only for convenience. And I'm all about convenience :) 

As for the accuracy, that might very well be. Reported metrics by Go multiplied by ~ 1.125 seems to give almost exactly what `top` and Activity Monitor reports, so I guess it's either some overhead of the application Go does not account for, or it is the OS allocating some overhead (?) which is not seen from inside of the app. I have frankly no idea 😅

What do you think regarding the default formatting?

**Comment by srenatus** [comment]
I think `alloc_bytes` would be a better name; or `total_alloc_bytes`. 

Re Default formatting, I'd keep it as it is right now. But I'm not having a strong opinion here. If we change the default, the name should perhaps include `_pretty`? 🤔 

Another thing, do we need to do something to ensure that the route won't clash with existing ones, or future ones, of the prom handler?

**Comment by anderseknert** [comment]
`alloc_bytes` is fine with me, although it sorta "conflicts" with the `pretty=true` option, or what do you think?

I suppose we could have one `/metrics/alloc_bytes` and another `/metrics/alloc_pretty`? 🤔

I have checked to make sure `/metrics` still works as before. The chance of Prometheus also adding an `/alloc` endpoint to scrape from should be slim, but if we're worried about it, we could use some other endpoint altogether. It does feel rather "metrics" oriented though.

**Comment by srenatus** [comment]
👍

**Comment by anderseknert** [comment]
@srenatus I'm sorry, was the thumb up about
> I suppose we could have one /metrics/alloc_bytes and another /metrics/alloc_pretty?

or that `/metrics` still works as expectd? 🙂

**Comment by srenatus** [comment]
The second. LGTM.

File: internal/prometheus/prometheus.go [hunk 1]
```diff
@@ -196,3 +200,31 @@ func (c *captureStatusResponseWriter) WriteHeader(statusCode int) {
...
+		bf /= 1000.0
+	}
+	return fmt.Sprintf("%.1fYiB", bf)
+}
+
+func allocHandler(rsp http.ResponseWriter, req *http.Request) {
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+
+	total := m.HeapInuse + m.StackInuse + m.MCacheInuse + m.MSpanInuse
```
Comment (ashutosh-narkar) [inline internal/prometheus/prometheus.go:219]: Can you help me understand why we these metrics and not something like `HeapAlloc`?

**Comment by ashutosh-narkar** [comment]
> I suppose we could have one `/metrics/alloc_bytes` and another `/metrics/alloc_pretty`?

I agree something like  `/metrics/alloc_bytes`  or `metrics/mem_stats` or `/metrics/alloc_mem_bytes` would be better.

File: internal/prometheus/prometheus.go (hunk 1 above)
Comment (anderseknert) [reply internal/prometheus/prometheus.go:219]: `Alloc`/`HeapAlloc` will show a lower number, moving the metrics reported here further away from what other tools report.

**Comment by anderseknert** [comment]
Alright, I'll update to use `alloc_bytes` tomorrow. Thanks to you both!

**Comment by srenatus** [comment]
I think there's another metric that could be useful for triangulating the memstats we're adding here, and what htop etc show: `process_resident_memory_bytes`. It's gathered by the prometheus process collector, and I think enabled by default. I'd expect that one to be closer to what htop presents (it's also [called RSS](https://github.com/prometheus/client_golang/blob/3d2cf0b338e19b0eaf277496058fc77bd3add440/prometheus/process_collector.go#L93-L96), like htop)

**Comment by anderseknert** [comment]
Thanks @srenatus.  I can definitely add that too. Do you know which of the [memstat](https://pkg.go.dev/runtime#MemStats) metrics that corresponds to?

**Comment by srenatus** [comment]
> Do you know which of the [memstat](https://pkg.go.dev/runtime#MemStats) metrics that corresponds to?

I don't think it corresponds to any: it's using /proc to find data about the _process'_ memory usage as seen from the kernel. But I just realised it's linux-only, so that might not be helpful here.

File: internal/prometheus/prometheus.go [hunk 2]
```diff
@@ -196,3 +200,31 @@ func (c *captureStatusResponseWriter) WriteHeader(statusCode int) {
...
+
+	var alloc string
+	if req.URL.Query().Get("pretty") == "true" {
+		alloc = prettyByteSize(total)
+	} else {
+		alloc = fmt.Sprintf("%d", total)
+	}
+
+	rsp.WriteHeader(200)
+	_, _ = rsp.Write([]byte(fmt.Sprintf("%v\n", alloc)))
```
Comment (srenatus) [inline internal/prometheus/prometheus.go:229]: Sorry for the last-minute nitpick:
```suggestion
	fmt.Fprintln(rsp, alloc)
```

**Review by srenatus — APPROVED** [review]

File: internal/prometheus/prometheus.go (hunk 2 above)
Comment (anderseknert) [reply internal/prometheus/prometheus.go]: Neat. Fixed, thanks!

**Review by charlieegan3 — APPROVED** [review]

**Review by srenatus — APPROVED** [review]

**Pull request merged**